File: app/src/main/java/com/sc/scapp/DownloadActivity.java
package com.sc.scapp;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.content.SharedPreferences;
import android.os.IBinder;
import android.os.Bundle;
import android.view.MenuItem;

import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Map;

public class DownloadActivity extends AppCompatActivity {

    private ArrayList<DownloadState> downloadStateList = new ArrayList<>();
    private DownloadAdapter downloadAdapter;
    private DownloadService.DownloadBinder downloadBinder;
    private ServiceConnection serviceConnection = new ServiceConnection() {
        @Override
        public void onServiceConnected(ComponentName name, IBinder service) {
            downloadBinder = (DownloadService.DownloadBinder) service;
            downloadBinder.setDownloadListener(downloadListener);

            SharedPreferences pref = getSharedPreferences(DownloadService.DOWNLOAD_SHARED_PREFERENCE_NAME, MODE_PRIVATE);
            Map<String, ?> prefMap = pref.getAll();
            for (String url : prefMap.keySet()) {
                String value = (String) prefMap.get(url);
                DownloadState downloadState = DownloadState.fromString(value);
                if (downloadState.getState() == DownloadState.DOWNLOADING && !downloadBinder.isDownloading(url)) {
                    downloadState.setState(DownloadState.DOWNLOAD_PAUSED);
                }
                downloadState.setUrl(url);
                downloadStateList.add(downloadState);
            }
            Collections.sort(downloadStateList);
            downloadAdapter.notifyDataSetChanged();
            downloadAdapter.setDownloadBinder(downloadBinder);
        }

        @Override
        public void onServiceDisconnected(ComponentName name) {

        }
    };
    private DownloadListener downloadListener = new DownloadListener() {
        @Override
        public void onProgress(String url, long downloadedLength, long contentLength) {
            for (int i = 0; i < downloadStateList.size(); i++) {
                DownloadState downloadState = downloadStateList.get(i);
                if (downloadState.getUrl().equals(url)) {
                    downloadState.setDownloadedLength(downloadedLength);
                    downloadState.setContentLength(contentLength);
                    downloadState.setState(DownloadState.DOWNLOADING);
                    downloadAdapter.notifyItemChanged(i);
                    break;
                }
            }
        }

        @Override
        public void onSuccess(String url) {
            for (int i = 0; i < downloadStateList.size(); i++) {
                DownloadState downloadState = downloadStateList.get(i);
                if (downloadState.getUrl().equals(url)) {
                    downloadState.setState(DownloadState.DOWNLOAD_SUCCESS);
                    downloadAdapter.notifyItemChanged(i);
                    break;
                }
            }
        }

        @Override
        public void onFailed(String url) {
            for (int i = 0; i < downloadStateList.size(); i++) {
                DownloadState downloadState = downloadStateList.get(i);
                if (downloadState.getUrl().equals(url)) {
                    downloadState.setState(DownloadState.DOWNLOAD_SUCCESS);
                    downloadAdapter.notifyItemChanged(i);
                    break;
                }
            }
        }

        @Override
        public void onPaused(String url) {
            for (int i = 0; i < downloadStateList.size(); i++) {
                DownloadState downloadState = downloadStateList.get(i);
                if (downloadState.getUrl().equals(url)) {
                    downloadState.setState(DownloadState.DOWNLOAD_PAUSED);
                    downloadAdapter.notifyItemChanged(i);
                    break;
                }
            }
        }

        @Override
        public void onCanceled(String url) {
            for (int i = 0; i < downloadStateList.size(); i++) {
                DownloadState downloadState = downloadStateList.get(i);
                if (downloadState.getUrl().equals(url)) {
                    //downloadState.setState(DownloadState.DOWNLOAD_CANCELED);
                    //downloadAdapter.notifyItemChanged(i);
                    downloadStateList.remove(i);
                    downloadAdapter.notifyItemRemoved(i);
                    break;
                }
            }
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_download);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null) {
            actionBar.setDisplayHomeAsUpEnabled(true);
        }

        RecyclerView recyclerView = findViewById(R.id.recycler_view);
        recyclerView.setLayoutManager(new LinearLayoutManager(DownloadActivity.this));
        downloadAdapter = new DownloadAdapter(downloadStateList);
        recyclerView.setAdapter(downloadAdapter);
        recyclerView.addItemDecoration(new DividerItemDecoration(this, DividerItemDecoration.VERTICAL));

        Intent bindIntent = new Intent(this, DownloadService.class);
        bindService(bindIntent, serviceConnection, BIND_AUTO_CREATE);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        downloadBinder.setDownloadListener(null);
        unbindService(serviceConnection);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                finish();
                break;
        }
        return true;
    }

    public static void actionStart(Context context) {
        Intent intent = new Intent(context, DownloadActivity.class);
        context.startActivity(intent);
    }
}

File: app/src/main/java/com/sc/scapp/DownloadActivity.kt
package com.sc.scapp

import android.content.*
import android.os.Bundle
import android.os.IBinder
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.sc.scapp.DownloadService.DownloadBinder
import java.util.*

class DownloadActivity : AppCompatActivity() {
    private val downloadStateList = ArrayList<DownloadState>()
    private lateinit var downloadAdapter: DownloadAdapter
    private lateinit var downloadBinder: DownloadBinder
    private val serviceConnection: ServiceConnection = object : ServiceConnection {
        override fun onServiceConnected(name: ComponentName, service: IBinder) {
            downloadBinder = service as DownloadBinder
            downloadBinder.setDownloadListener(downloadListener)
            val pref = getSharedPreferences(DownloadService.DOWNLOAD_SHARED_PREFERENCE_NAME, MODE_PRIVATE)
            val prefMap = pref.all
            for (url in prefMap.keys) {
                val value = prefMap[url] as String?
                val downloadState = DownloadState.fromString(value)
                if (downloadState.state == DownloadState.DOWNLOADING && !downloadBinder.isDownloading(url)) {
                    downloadState.state = DownloadState.DOWNLOAD_PAUSED
                }
                downloadState.url = url
                downloadStateList.add(downloadState)
            }
            downloadStateList.sort()
            downloadAdapter.notifyDataSetChanged()
            downloadAdapter.setDownloadBinder(downloadBinder)
        }

        override fun onServiceDisconnected(name: ComponentName) {}
    }
    private val downloadListener: DownloadListener = object : DownloadListener {
        override fun onProgress(url: String, downloadedLength: Long, contentLength: Long) {
            for (i in downloadStateList.indices) {
                val downloadState = downloadStateList[i]
                if (downloadState.url == url) {
                    downloadState.downloadedLength = downloadedLength
                    downloadState.contentLength = contentLength
                    downloadState.state = DownloadState.DOWNLOADING
                    downloadAdapter.notifyItemChanged(i)
                    break
                }
            }
        }

        override fun onSuccess(url: String) {
            for (i in downloadStateList.indices) {
                val downloadState = downloadStateList[i]
                if (downloadState.url == url) {
                    downloadState.state = DownloadState.DOWNLOAD_SUCCESS
                    downloadAdapter.notifyItemChanged(i)
                    break
                }
            }
        }

        override fun onFailed(url: String) {
            for (i in downloadStateList.indices) {
                val downloadState = downloadStateList[i]
                if (downloadState.url == url) {
                    downloadState.state = DownloadState.DOWNLOAD_SUCCESS
                    downloadAdapter.notifyItemChanged(i)
                    break
                }
            }
        }

        override fun onPaused(url: String) {
            for (i in downloadStateList.indices) {
                val downloadState = downloadStateList[i]
                if (downloadState.url == url) {
                    downloadState.state = DownloadState.DOWNLOAD_PAUSED
                    downloadAdapter.notifyItemChanged(i)
                    break
                }
            }
        }

        override fun onCanceled(url: String) {
            for (i in downloadStateList.indices) {
                val downloadState = downloadStateList[i]
                if (downloadState.url == url) {
                    //downloadState.setState(DownloadState.DOWNLOAD_CANCELED);
                    //downloadAdapter.notifyItemChanged(i);
                    downloadStateList.removeAt(i)
                    downloadAdapter.notifyItemRemoved(i)
                    break
                }
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_download)
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
        val recyclerView = findViewById<RecyclerView>(R.id.recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(this@DownloadActivity)
        downloadAdapter = DownloadAdapter(downloadStateList)
        recyclerView.adapter = downloadAdapter
        recyclerView.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
        val bindIntent = Intent(this, DownloadService::class.java)
        bindService(bindIntent, serviceConnection, BIND_AUTO_CREATE)
    }

    override fun onDestroy() {
        super.onDestroy()
        downloadBinder.setDownloadListener(null)
        unbindService(serviceConnection)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> finish()
        }
        return true
    }

    companion object {
        @JvmStatic
        fun actionStart(context: Context) {
            val intent = Intent(context, DownloadActivity::class.java)
            context.startActivity(intent)
        }
    }
}